hare: 0.24.2 -> 0.25.2

liljamo opened the pull request:
https://harelang.org/blog/2025-06-21-hare-0.25.2-released/
https://git.sr.ht/~sircmpwn/harec/refs/0.25.2
https://git.sr.ht/~sircmpwn/hare/refs/0.25.2

Major question marks include the new "hare tool" subcommand and how it works, see the blog post for details.
Might be possible to do this update first, and then figure out the tool stuff when we package [hare-update](https://git.sr.ht/~sircmpwn/hare-update)?

cc @onemoresuza 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

liljamo committed:
harec: 0.24.2 -> 0.25.2

liljamo committed:
hare: 0.24.2 -> 0.25.2

onemoresuza commented:
Hi, @liljamo.

Perhaps the usage of the `MakeFlags` or the `makeFlagsArray` env var may be a more nixy way of passing the C compiler flag.

Regarding both this PR and [hare's](https://github.com/NixOS/nixpkgs/pull/425176), I'd recommend doing as it was done on [#325334](https://github.com/NixOS/nixpkgs/pull/325334/), i. e., one PR with one commit for every updated package, since by doing so:

1. We avoid having incompatible versions of hare and harec, for instance, in the chance that one PR is merged and the other not; and
2. We ensure that the Hare packages are working with the new version of Hare --- since Hare's ecosystem doesn't have that much packages on nixpkgs, I've found it better to update them on the same PR

liljamo committed:
hareThirdParty.hare-toml: 0.1.2 -> 0.2.0

liljamo commented:
> Perhaps the usage of the `MakeFlags` or the `makeFlagsArray` env var may be a more nixy way of passing the C compiler flag.

I tried using those both, but I couldn't find a way to extend the existing `CFLAGS` array with them. Could be that I'm missing something and there is a way?
Another option would be to hardcode the flags from `config.mk` to the `makeFlags` array, I guess. I tried searching around in nixpkgs, and found a mix of both hardcoding and doing the echo thing that I came up with.
I don't really have a hard opinion on this, since I'm a new contributor here.


> one PR with one commit for every updated package

I've moved the hare bump here and closed the other PR. I also moved the note about the `hare tool` command, do you have any takes on that? Whether to include it in this here update, or do it separately?

As for the packages that use `hareHook`, I'll see about updating those that have already updated to 0.25.2, but we're probably going to be waiting a while, since 0.25.2 had breaking changes that affect all hare programs.

ilylily commented:
rather than ignoring the warning with compiler flags, it might be better to apply commit [`31df27c26e026f3d934759f0e7b3ff105e7bf74c`](https://git.sr.ht/~sircmpwn/harec/commit/31df27c26e026f3d934759f0e7b3ff105e7bf74c), as a patch. this commit silences a few gcc warnings with very minor code changes

liljamo commented:
> rather than ignoring the warning with compiler flags, it might be better to apply commit [`31df27c26e026f3d934759f0e7b3ff105e7bf74c`](https://git.sr.ht/~sircmpwn/harec/commit/31df27c26e026f3d934759f0e7b3ff105e7bf74c), as a patch. this commit silences a few gcc warnings with very minor code changes

Hey, thanks for the note, applied!

As for the rest of this PR, I'm hoping to have time soon to check if all(/most) hare programs in nixpkgs have updated to address the breaking changes.
